analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs

```diff
@@ -99,6 +99,20 @@ private ProgramState()
     public ProgramState SetSymbolConstraint(ISymbol symbol, SymbolicConstraint constraint) =>
         SetSymbolValue(symbol, (this[symbol] ?? SymbolicValue.Empty).WithConstraint(constraint));
 
+    public bool HasConstraint(IOperation operation, SymbolicConstraint constraint) =>
+        operation is not null
```

Comment:
Asking a question for something (`operation` and `symbol` below) that is null still sounds weird to me.

Can you take a look at existing usages and identify how many null-checks we can remove from existing usages? By now, we should already have a representative sample. Also taking into account situations that previously used the manual `state[something]?.HasConstraint` that will be refactored to this new. Ideally, as part of this PR to prove the point.

Reply:
I found 3 cases: [#1](https://github.com/SonarSource/sonar-dotnet/blob/f4cfe1cfe46dc3df40a73434e62b605932234c2c/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs#L130-L131) [#2](https://github.com/SonarSource/sonar-dotnet/blob/f4cfe1cfe46dc3df40a73434e62b605932234c2c/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs#L208-L210) [#3](https://github.com/SonarSource/sonar-dotnet/blob/f4cfe1cfe46dc3df40a73434e62b605932234c2c/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs#L219)

A different approach would be an extension method on `IOperation`:
`operation.HasConstraint(state, BoolConstraint.True)`
In that case, the argument for making it not null-safe would be easier since adding a null check via `?.` on the caller side would be trivial.

Reply:
#1 - ok
#2 - this seems to me that it would change the actual behavior.
#3 - ok
And I think we had at least one NREs around this due to static properties.  So OK, let's keep it null tolerant.